src/components/valences-test/ValencesTest.tsx

```diff
@@ -16,202 +15,190 @@ import TestResults from "../test-results/TestResults";
   data: IElement;
 }
 
-interface IValencesTestState {
-  questions: IValencesTestQuestionCard[];
-  right: IValencesTestQuestionCard[];
-  wrong: IValencesTestQuestionCard[];
+interface ValencesTestState {
+  questions: ValencesTestQuestionCard[];
+  right: ValencesTestQuestionCard[];
+  wrong: ValencesTestQuestionCard[];
 }
 
-@autobind
-class ValencesTest extends React.Component<Props> {
-  public state: IValencesTestState = {
-    questions: [],
-    right: [],
-    wrong: [],
-  };
-
-  private settings: IValencesTestSettings = getValencesTestSettings();
+function ValencesTest() {
+  const settings = React.useRef(getValencesTestSettings() || []).current;
 
-  public componentDidMount() {
-    this.createTestQuestions();
+  function createAnswer(answer: string, right = false): Answer {
+    return {
+      answer,
+      right,
+    };
   }
 
-  public render() {
-    const { questions, wrong, right } = this.state;
-    const hasQuestions = !!questions.length;
+  function createQuestionAnswers(element: IElement): Answer[] {
+    const rightAnswer = createAnswer(element.valency, true);
+    const wrongAnswerPool = shuffle(element.wrongValences)
+      .map((wrongValency) => createAnswer(wrongValency))
+      .slice(0, 3);
 
-    return (
-      <div className="valences-test">
-        <Navbar
-          title={i18n("valences_test")}
-          onBackButtonClick={this.onNavbarBackButtonClick}
-        />
+    return shuffle([rightAnswer, ...wrongAnswerPool]);
+  }
 
-        {hasQuestions && (
-          <div className="valences-test__test">
-            <QuestionsTest
-              title={i18n("select_valence")}
-              questions={questions}
-              // @ts-ignore Fix types
-              onQuestionAnswer={this.onQuestionAnswer}
-            />
-          </div>
-        )}
-
-        {!hasQuestions && (
-          <div className="valences-test__result">
-            <Card className="valences-test__result-card">
-              <TestResults
-                gaTestName="Valences Test"
-                wrongAnswers={wrong.length}
-                rightAnswers={right.length}
-                onRepeat={this.repeatTest}
-                onRepeatWrongAnswers={this.repeatWrongAnswers}
-              />
-            </Card>
-          </div>
-        )}
-      </div>
-    );
+  function createQuestion(element: IElement): ValencesTestQuestionCard {
+    return {
+      answers: createQuestionAnswers(element),
+      data: element,
+      question: element.symbol,
+      questionClass: `valences-test__question element ${element.group}`,
+    };
   }
 
-  private onNavbarBackButtonClick() {
-    const { history } = this.props;
+  function createTestQuestions() {
+    const questions = settings!
+      .elements!.filter((element) => element.enabled)
+      .map((element) => ElementManager.getElement(element.atomic))
+      .map((element) => createQuestion(element!));
 
-    history.push(TEST_SELECTION);
+    return shuffle(questions);
   }
```

Comment:
This is the only `createX` function that actually needs to be defined _in_ the render function and only because it needs access to `settings`.  If it accepted `settings` as a parameter, the could all move outside, e.g.

```ts
function createAnswer(answer: string, right = false): Answer {
  return {
    answer,
    right,
  };
}

function createQuestionAnswers(element: IElement): Answer[] {
  const rightAnswer = createAnswer(element.valency, true);
  const wrongAnswerPool = shuffle(element.wrongValences)
    .map((wrongValency) => createAnswer(wrongValency))
    .slice(0, 3);

  return shuffle([rightAnswer, ...wrongAnswerPool]);
}

function createQuestion(element: IElement): ValencesTestQuestionCard {
  return {
    answers: createQuestionAnswers(element),
    data: element,
    question: element.symbol,
    questionClass: `valences-test__question element ${element.group}`,
  };
}

function createTestQuestions(settings: IValencesTestSettings) {
  const questions = settings
    .elements!.filter((element) => element.enabled)
    .map((element) => ElementManager.getElement(element.atomic))
    .map((element) => createQuestion(element!));

  return shuffle(questions);
}

function ValencesTest() {
  // ...
}
```

and remember to call it as `createTestQuestions(settings)`.

Reply:
Yes, we can totally do that too 🙂